poller-service.py

```diff
@@ -0,0 +1,286 @@
+             'AND IS_FREE_LOCK(CONCAT("queued.", device_id))                     '
+             'AND ( last_poll_attempted < DATE_SUB(NOW(), INTERVAL {2} SECOND )  '
+             '  OR last_poll_attempted IS NULL )                                 '
+             '{3}                                                                '
+             'ORDER BY next_poll asc                                             ').format(poll_frequency,
+                                                                                           discover_frequency,
+                                                                                           down_retry,
+                                                                                           poller_group)
+
+threads = 0
+next_update = datetime.now() + timedelta(minutes=1)
+devices_scanned = 0
+
+while True:
+    cur_threads = threading.active_count()
+    if cur_threads != threads:
+        threads = cur_threads
+        log.debug('DEBUG: {} threads currently active'.format(threads))
+
+    if next_update < datetime.now():
+        seconds_taken = (datetime.now() - (next_update - timedelta(minutes=1))).seconds
+        update_query = ('INSERT INTO pollers(poller_name,     '
+                        '                    last_polled,     '
+                        '                    devices,         '
+                        '                    time_taken)      '
+                        '  values("{0}", NOW(), "{1}", "{2}") '
+                        'ON DUPLICATE KEY UPDATE              '
+                        '  last_polled=values(last_polled),   '
+                        '  devices=values(devices),           '
+                        '  time_taken=values(time_taken)      ').format(config['distributed_poller_name'].strip(),
+                                                                        devices_scanned,
+                                                                        seconds_taken)
+        try:
+            cursor.execute(update_query)
+        except:
+            log.critical('ERROR: MySQL query error. Is your schema up to date?')
+            sys.exit(2)
+        cursor.fetchall()
+        log.info('INFO: {} devices scanned in the last 5 minutes'.format(devices_scanned))
+        devices_scanned = 0
+        next_update = datetime.now() + timedelta(minutes=1)
+
+    while threading.active_count() >= amount_of_workers:
+        time.sleep(.5)
+
+    try:
+        cursor.execute(dev_query)
+    except:
+        log.critical('ERROR: MySQL query error. Is your schema up to date?')
+        sys.exit(2)
+
+    devices = cursor.fetchall()
+    for device_id, status, next_poll, next_discovery in devices:
+        # add queue lock, so we lock the next device against any other pollers
+        # if this fails, the device is locked by another poller already
+        if not getLock('queued.{}'.format(device_id)):
+            time.sleep(.5)
+            continue
```

Comment:
We need to get more than one, just in case there is a close race with another poller and we fail to get a lock on the device. This allows us to continue to the next device.

Something like limit=5 might be a good idea though. Would most likely be enough to find one to get a lock on, and if all 5 fail it will query again anyway. And it should reduce the sql load on environments with lots of devices.

Other things I might want to do to reduce load would be to add an index on last_poll_attempted. I'll test that shortly.